python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,745 @@
+    return IntegerLiteral(value)
+
+
+@literal.register
+def _(value: float):
+    """
+    Upgrade to double if python float is outside the JAVA_MIN_FLOAT and JAVA_MAX_FLOAT
+    """
+    if value < JAVA_MIN_FLOAT or value > JAVA_MAX_FLOAT:
+        return DoubleLiteral(value)
+    return FloatLiteral(value)
+
+
+@literal.register
+def _(value: str):
+    return StringLiteral(value)
+
+
+@literal.register
+def _(value: uuid.UUID):
+    return UUIDLiteral(value)
+
+
+@literal.register
+def _(value: bytes):
+    return FixedLiteral(value)
+
+
+@literal.register
+def _(value: bytearray):
+    return BinaryLiteral(value)
+
+
+@literal.register
+def _(value: Decimal):
+    return DecimalLiteral(value)
+
+
+class BaseLiteral(Literal):
+    """Base literal which has a value and can be converted between types"""
+
+    def __init__(self, repr_string: str, value):
+        self._repr_string = repr_string
+        self.value = value
+        self.byte_buffer = None
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def __eq__(self, other):
+        if id(self) == id(other):
+            return True
+        elif other is None or not isinstance(other, BaseLiteral):
+            return False
+
+        return self.value == other.value
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        return self._repr_string
+
+    def __str__(self):
+        return str(self.value)
+
+    def to_byte_buffer(self):
+        if self.byte_buffer is None:
+            # TODO need byte buffer conversion
+            raise NotImplementedError()
+
+        return self.byte_buffer
+
+
+class ComparableLiteral(BaseLiteral):
+    def __init__(self, repr_string: str, value):
+        super(ComparableLiteral, self).__init__(repr_string, value)
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def __eq__(self, other):
+        return self.value == other.value
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __lt__(self, other):
+        if self.value is None:
+            return True
+
+        if other is None or other.value is None:
+            return False
+
+        return self.value < other.value
+
+    def __gt__(self, other):
+        if self.value is None:
+            return False
+
+        if other is None or other.value is None:
+            return True
+
+        return self.value > other.value
+
+    def __le__(self, other):
+        if self.value is None:
+            return True
+
+        if other is None or other.value is None:
+            return False
+
+        return self.value <= other.value
+
+    def __ge__(self, other):
+        if self.value is None:
+            return False
+
+        if other is None or other.value is None:
+            return True
+
+        return self.value >= other.value
+
+
+class AboveMax(Literal, Singleton):
+    _instance = None
+
+    def __init__(self):
+        self._initialized = True
```

Comment:
It's not clear to me why we are setting this flag here. I don't think this is the same situation as this was used in `types.py`. There's no check for whether this flag is set here to avoid calling the super init so if this flag really is needed I think we need the check too. If we don't need the flag, I think this init overall can be gotten rid of, which I think makes sense right now